src/clr-angular/forms/combobox/option-items.ts

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2016-2020 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { NgForOf, NgForOfContext } from '@angular/common';
+import {
+  Directive,
+  DoCheck,
+  Input,
+  IterableDiffer,
+  IterableDiffers,
+  TemplateRef,
+  TrackByFunction,
+  ViewContainerRef,
+  OnDestroy,
+} from '@angular/core';
+
+import { OptionSelectionService } from './providers/option-selection.service';
+import { Subscription } from 'rxjs';
+import { ClrPopoverPositionService } from '../../utils/popover/providers/popover-position.service';
+
+@Directive({
+  selector: '[clrOptionItems][clrOptionItemsOf]',
+})
+export class ClrOptionItems<T> implements DoCheck, OnDestroy {
+  private iterableProxy: NgForOf<T>;
+  private _rawItems: T[];
+  private filteredItems: T[];
+  private subscriptions: Subscription[] = [];
+  private filter: string;
+  private _filterField: string;
+  private differ: IterableDiffer<T> | null = null;
+
+  @Input('clrOptionItemsOf')
+  public set rawItems(items: T[]) {
+    this._rawItems = items ? items : [];
+  }
+
+  @Input('clrOptionItemsTrackBy')
+  set trackBy(value: TrackByFunction<T>) {
+    this.iterableProxy.ngForTrackBy = value;
+  }
+
+  @Input('clrOptionItemsField')
+  public set field(field: string) {
+    this._filterField = field;
+    this.optionService.displayField = field;
+  }
+
+  constructor(
+    public template: TemplateRef<NgForOfContext<T>>,
+    private differs: IterableDiffers,
+    private optionService: OptionSelectionService<T>,
+    private positionService: ClrPopoverPositionService,
+    private vcr: ViewContainerRef
+  ) {
+    this.iterableProxy = new NgForOf<T>(this.vcr, this.template, this.differs);
+    this.subscriptions.push(
+      optionService.inputChanged.subscribe(filter => {
+        this.filter = filter;
+      })
+    );
+  }
+
```

Comment:
I'm concerned that this code will run alot and for large arrays of items could be expensive for apps to run. It looks like this was patterned after the `*clrDgItems` smart iterator in the datagrid and one of the reasons it works is that the only time code is executed is when the differ has changes for the items provided to it. 

Here, we are creating filteredItems everytime change detection is run. Given that this is an input its alot and this is different from the datagrid apporach too, its only an input on the row so change detection isn't nearly as chatty as it is with all the events that trigger ngDoCheck when I run this code now. 

What about this, move lines #68-#82 into a new private function. In my test I called it `updateFilteredItems(): void {...}`. Then I only need to call it in the setter for `@Input('clrOptionItemsOf')` and in the `optionService.inputChanged` subscription. 

I think this limits the work that needs to be done when ngDoCheck is executed.

Reply:
Tested your suggestion. It really makes difference. Thanks!
I will only leave line 82 (the call to `iterableProxy.ngDoCheck()`) back, so that I'll not be breaking the natural `ngForOf` chain, which I am building upon in this directive.